design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,542 @@
+other parameters are recommended, if available, to assist the developer with diagnosing the error.
+ 
+```cpp
+[[nodiscard]] IXlangErrorInfo* xlang_originate_error(
+    xlang_result error,
+    xlang_string message,
+    xlang_string projectionIdentifier,
+    xlang_string languageError,
+    IXlangErrorExecutionTrace* executionTrace,
+    IUnknown* languageInformation)
+```
+
+If the language error is a propagation of a previous xlang error (i.e. the component called another xlang
+projected component which returned an error), then the associated error info should be retrieved.
+The retrieved xlang error info and the associated xlang error should be used as the error instead of
+translating the language error and creating a new xlang error info. This will allow the identity
+of the error to be independent of the translations it may go through as it propagates through
+various languages and for it to be defined as what the initial projection which returned the error
+determined it was equivalent to. Diagnostic information collected by the language should still be
+added to the error info as it can provide useful information in the case the error becomes a failure.
+This can be done by a call to `PropagateError` which will add the provided information to the
+error info.
+
+`PropagateError` will be defined in the `IXlangErrorInfo` interface and will take the following parameters:
+- a `xlang_string` indicating an identifier for the language
+- a `xlang_string` with the language specific error
+- a `IXlangErrorExecutionTrace` with the execution trace if available
+- an `IUnknown` pointer with any additional language specific information (can be null)
+
+```cpp
+void PropagateError(
+    xlang_string projectionIdentifier,
+    xlang_string languageError,
+    IXlangErrorExecutionTrace* executionTrace,
+    IUnknown* languageInformation)
+```
+
+After calling xlang's origination / propagation API, the projection should return the error info as the
+result so that the consuming projection can receive it and propagate it to the language it projects.
+
+## Projection identifiers
+
+xlang recommends projections to choose a unique projection identifier and provide it as part of the
+error origination and propagation APIs. This identifier will help developers know which language the
+error came from as it may not be obvious from the language error and the execution trace. It will
+also help tooling determine whether the diagnostic information is from a projection that it recognizes
+and can light up additional features such as converting the execution trace to something developer friendly.
+
+Given that projections may evolve over time and change what they store as part of the diagnostic
+information, the identifier provided should also have a version number. Due to this, xlang recommends
+the identifier be in the form of <unique_projection_identifer>_<version_number>.
+
+## Execution traces
```

Comment:
per discussion, this is great stuff, but best factored out into a separate design note.  The IUnknown executionTrace param above makes a neat firewall for discussing these topics separately, and this one was just introduced.

Reply:
Removed it from this design note for now and marked this section as in progress.  Will start a new PR with this section.